### DISCUSSION TITLE
Representing spheres

### DISCUSSION THREAD

**Discussion by A-Walrus** [hannobraun/fornjot#1720] [General]
I understand that this is not planned for the current milestone, but still think it's important to discuss. Currently surface geometry is represented as
```rust
/// The geometry that defines a surface
#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash, Ord, PartialOrd)]
pub struct SurfaceGeometry {
    /// The u-axis of the surface
    pub u: GlobalPath,

    /// The v-axis of the surface
    pub v: Vector<3>,
}
```
This approach cannot handle something which is curved in multiple directions like a sphere. Are there any plans/ideas in this regard?

**Comment by hannobraun** [comment]
I only have vague ideas at this point. It's something I've had in the back of my head for a while, but I haven't really looked into developing solutions.

Some thoughts:
- Could we represent a sphere by also using a `GlobalPath` for the `v` axis? That would lead to a weird, self-intersecting surface, but maybe the coordinate system would work out correctly? (Also see #997.)
- We can convert `SurfaceGeometry` into an enum, and only make the current representation one of several that are possible. Then we could add a `Sphere` variant.
- At some point, we probably want to go for [NURBS](https://en.wikipedia.org/wiki/Non-uniform_rational_B-spline), or something similarly powerful like [T-splines](https://en.wikipedia.org/wiki/T-spline). This will undoubtedly bring many complications and be a large project. I don't know when that will make sense. But nothing is stopping us from converting `SurfaceGeometry` into an enum and add a `Nurbs` variant, to slowly start developing this. Supporting spheres could be a first milestone.

So overall, I don't think I have many helpful things to say, and I don't expect that I will be working on this any time soon. But if you (or someone else) wants to experiment, I'd be happy to assist as best I can!

**Reply by axelkar** [reply]
Hi! Is there development on this?

**Reply by hannobraun** [reply]
Hey @axelkar! Lots of development, but no actual progress on this specific issue.

A lot of the work I've done since this discussion was not directly related to improving the core geometry representation. More recently, I've been focusing on just that, but a lot of groundwork was necessary first. Most recently, I've been questioning whether the current architecture is suited to support the expanded geometric primitives that I was working towards, and I've been experimenting with different approaches since then.

If these experiments are successful, then things like representing spheres could become much easier to implement. But at this point, I really can't make any firm predictions.
